Product Gallery Thumbnails: Add View all overlay

danieldudzic opened the pull request:
This PR adds the "View all" overlay to the last thumbnail which triggers the lightbox/dialog (which will display all the product gallery images).

<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

## What

Fixes #11083

## Why

The "View all" overlay provides the user with feedback about the total number of available product images in the gallery (and offer a way to view them). 

## Testing Instructions

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. With a block theme, open `Appearance > Editor`.
2. Open `Templates > Single Product`.
3. Select `Convert to Blocks` if you are using the Classic Templates.
4. Add the `Product Gallery` block.
5. Ensure the Thumbnails are enabled.
6. Open a product page on the frontend **with multiple product images** (more than `Number of Thumbnails` setting) and test the frontend. Make sure the last thumbnail displays an overlay that opens the Dialog on click.
7. Make sure the overlay displays the correct number of remaining thumbnails.
8. Make sure there are no regressions and the Dialog can still be triggered by clicking the Large Image.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

## Screenshots or screencast

<!-- Any screenshots of UI changes will be helpful to include here. Leave blank if not applicable. -->

| Before | After |
| ------ | ----- |
|<img width="539" alt="V-Neck_T-Shirt_–_newproductgallery" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/7c2b9764-6bb9-462d-8d21-bcedc469bcd4">|<img width="543" alt="V-Neck_T-Shirt_–_newproductgallery" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/f28faeab-1d81-4b6d-86a5-860c3af4cb85">|

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental
* [ ] N/A

## Checklist

Required:
* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [x] This PR is assigned to a milestone.

Conditional:
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog
<!-- Provide a brief, descriptive summary of the changes in this PR. Include potential impacts on different parts of the product. Example: "Updated the checkout process to streamline the experience for users and reduce the number of steps." -->

> Product Gallery Thumbnails: Add a View all overlay.

danieldudzic committed:
Product Gallery Thumbnails: Add View All link to the last thumbnail (non-interactive)

danieldudzic committed:
Product Gallery Thumbnails: Add interactivity to the View All overlay

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

gigitux commented on src/BlockTypes/ProductGalleryThumbnails.php:
```diff
@@ -40,6 +40,38 @@ protected function get_block_type_uses_context() {
+	 * @return string
+	 */
+	protected function generate_view_all_html( $remaining_thumbnails_count ) {
+		return '<div class="wc-block-product-gallery-thumbnails__thumbnail__overlay wc-block-product-gallery-dialog-on-click" data-wc-on--click="actions.woocommerce.handleClick"><span class="wc-block-product-gallery-thumbnails__thumbnail__remaining-thumbnails-count wc-block-product-gallery-dialog-on-click">+' . esc_html( $remaining_thumbnails_count ) . '</span><span class="wc-block-product-gallery-thum ...
```
I think that it is more readable if we use `sprintf` or `strtr`.

gigitux reviewed (changes requested):
Thanks for working on this issue! I added a comment that I think that it makes sense to address! Also, given that the PR is very small, doesn't it make sense to fix #11099 and #11100 in this PR?

danieldudzic committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/11083-product-gallery-thumbnails-add-view-all

roykho commented:
Bumping this PR to the next release.

danieldudzic committed:
Product Gallery Thumbnails: Refactor View all html to make it more readable

danieldudzic committed:
Product Gallery Thumbnails: Fix #11100 - Load all thumbnails and hide the View all overlay when in Dialog

danieldudzic committed:
Product Gallery Thumbnails: Fix #11099 - Enable the dialog for the View all thumbnails overlay even when the 'Full-screen when clicked' setting is disabled

danieldudzic commented:
> Thanks for working on this issue! I added a comment that I think that it makes sense to address! Also, given that the PR is very small, doesn't it make sense to fix #11099 and #11100 in this PR?

Thank you for your feedback. Both issues have been fixed in subsequent commits :)

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

gigitux commented on src/BlockTypes/ProductGalleryLargeImage.php:
```diff
@@ -125,7 +125,7 @@ private function get_main_images_html( $context, $product_id ) {
 
 		if ( $context['fullScreenOnClick'] ) {
-			$attributes['class'] .= ' wc-block-woocommerce-product-gallery-large-image__image--full-screen-on-click';
+			$attributes['class'] .= ' wc-block-woocommerce-product-gallery-large-image__image--full-screen-on-click wc-block-product-gallery-dialog-on-click';
```
Is there a reason for this change?

danieldudzic replied:
Yes, the reason for this change is: that the `View All` thumbnails overlay is not an IMG and we still want it to trigger the opening of the Dialog.

gigitux replied:
We should create another action for the View All link. In this way, we can avoid the check on the CSS class. Does it make sense?

danieldudzic replied:
It does make sense! I'm open to that for sure.

However, semantically it also makes sense to me to have the same "action" (opening the Dialog) be handled by the same action.

Additionally, `tagName === 'IMG'` seems to be a bit too broad and narrowing it down will make it less fragile to potential conflicts. For example we have the ability to stack/add blocks on top of the Large Image. With the `tagName === 'IMG' check, if you add an image, then the Dialog will be triggered also when that "new" image is clicked.

What do you think?

gigitux replied:
> However, semantically it also makes sense to me to have the same "action" (opening the Dialog) be handled by the same action.

Not really. You can image them as different event handlers for different events. One is the click in the large image and the other one is the click on the View All thumbnails.

> Additionally, tagName === 'IMG' seems to be a bit too broad and narrowing it down will make it less fragile to potential conflicts. For example we have the ability to stack/add blocks on top of the Large Image. 

True, but I think that it is very uncommon use case. For now, I would keep it as it is. However, I don't have a strong opinion, so feel free to keep the check on the class if you think that it makes sense!

danieldudzic replied:
@roykho Do you have any opinion on this?

gigitux commented on assets/js/blocks/product-gallery/inner-blocks/product-gallery-large-image/frontend.tsx:
```diff
@@ -72,7 +72,11 @@ interactivityStore(
 				} ) => {
-					if ( ( event.target as HTMLElement ).tagName === 'IMG' ) {
+					if (
+						( event.target as HTMLElement ).classList.contains(
+							'wc-block-product-gallery-dialog-on-click'
+						)
```
Is there a reason for this change?

danieldudzic replied:
Yes, the reason for this change is: that the `View All` thumbnails overlay is not an IMG and we still want it to trigger the opening of the Dialog.

gigitux replied:
We should create another action for the View All link. In this way, we can avoid the check on the CSS class. Does it make sense?

danieldudzic replied:
https://github.com/woocommerce/woocommerce-blocks/pull/11087#discussion_r1355718444

roykho replied:
I feel like there should be a separate action "view all" for the product-gallery-thumbnails because we're in the context of thumbnails and not the large image right? Or am I missing something?

danieldudzic replied:
To me it's essetially the same context, as both do the same thing: Open the Lightbox (Dialog). No matter about the component that's being clicked.

roykho replied:
In the future when another developer looks into an issue with the gallery thumbnails, instinctively they will look into the product-gallery-thumbnails section as this pertains to that. It might confuse them, the code lives in the product-gallery-large-image. That's the only concern I had on that.

Aside: How come we need to check for the class onclick? If the action filter is named specific to opening the lightbox, we can just open it when click without first checking for specific class name and we can remove all these https://github.com/woocommerce/woocommerce-blocks/pull/11087/files#diff-865f98c619d1ca4f27d026d7c9bdd132a9d049158ff3b82701ed2e0426b42fc9R52-R53

danieldudzic replied:
> In the future when another developer looks into an issue with the gallery thumbnails, instinctively they will look into the product-gallery-thumbnails section as this pertains to that. It might confuse them, the code lives in the product-gallery-large-image. 

This is very valid. If this action is supposed to be shared among components it should live on the parent, Product Gallery.

> How come we need to check for the class onclick? If the action filter is named specific to opening the lightbox, we can just open it when click without first checking for specific class name and we can remove all these

So:
- The original action had an `IMG` check so it wouldn't get triggered by clicking on the Next/Previous buttons, Sale badge etc.
- To keep the code DRY I wanted to reuse the same Dialog opening action for the `View All` overlay.
- To achieve that I needed to remove the `IMG` check as the overlay is not an image.
- I have replaced it with a class check for better control and to enable it for both the Large Image and `View All` overlay.

gigitux commented on src/BlockTypes/ProductGalleryThumbnails.php:
```diff
@@ -37,7 +37,48 @@ protected function get_block_type_style() {
+	 *
+	 * @return string
+	 */
+	protected function generate_view_all_html( $remaining_thumbnails_count ) {
+		$view_all_html = '<div class="wc-block-product-gallery-thumbnails__thumbnail__overlay wc-block-product-gallery-dialog-on-click" data-wc-on--click="actions.woocommerce.handleClick">'
+			. '<span class="wc-block-product-gallery-thumbnails__thumbnail__remaining-thumbnails-count wc-block-product-gallery-dialog-on-click">+%1$s</span>'
+			. '<span class="wc-block-product-gallery-thumbnails__thumbnail__view-all wc-block-product-gallery-dialog-on-click">%2$s</span>'
+			. '</div>';
+
+		return sprintf(
+			$view_all_html,
+			esc_html( $remaining_thumbnails_count ),
+			__( 'View all', 'woo-gutenberg-products-block' )
+		);
+	}
```
Is there a specific reason for this concatenation? If not, I think that avoiding the concatenation improves the readability of the code. What do you think?

danieldudzic replied:
Fair point. Removed ✅

gigitux reviewed (commented):
Thanks for addressing the feedback! Overall the PR looks great. I left some questions, could you take a look? Also, I noticed that the E2E tests are failing. Could you check if your PR introduces some regression?

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Product Gallery Thumbnails: Remove unnecessary concatenation from the View all html

danieldudzic committed:
Merge branch 'fix/11083-product-gallery-thumbnails-add-view-all' of https://github.com/woocommerce/woocommerce-blocks into fix/11083-product-gallery-thumbnails-add-view-all

roykho commented:
Some notes from a brief scan of the functionality and code:

> The Dialog opening currently only works with the Full-screen when clicked setting enabled. This will be addressed in a separate PR.

I am able to get to the dialog box with the setting enabled and disabled. Not sure if anything changed since you wrote that note.

> The "View all" overlay should not be present in the Dialog, and all Thumbnails should be displayed. This will be addressed in a separate PR.

I am not seeing the "View All" in the dialog box and I am seeing all the thumbnails. Again, I am not sure if this has already been fixed in this PR since you wrote the note.

danieldudzic commented:
> > The Dialog opening currently only works with the Full-screen when clicked setting enabled. This will be addressed in a separate PR.
> 
> I am able to get to the dialog box with the setting enabled and disabled. Not sure if anything changed since you wrote that note.

I have now enabled that in https://github.com/woocommerce/woocommerce-blocks/pull/11087/commits/bb719e8c6af7c7fb9b5be2bf25a096298bc3cb79

> > The "View all" overlay should not be present in the Dialog, and all Thumbnails should be displayed. This will be addressed in a separate PR.

> I am not seeing the "View All" in the dialog box and I am seeing all the thumbnails. Again, I am not sure if this has already been fixed in this PR since you wrote the note.

I have now addressed this in https://github.com/woocommerce/woocommerce-blocks/pull/11087/commits/624035d36ffa01a0182c40024c605bc28d2423c8

Thank you for pointing this out. I have now updated the PR description.

roykho commented on src/BlockTypes/ProductGalleryThumbnails.php:
```diff
@@ -37,7 +37,48 @@ protected function get_block_type_style() {
+		return sprintf(
+			$view_all_html,
+			esc_html( $remaining_thumbnails_count ),
+			__( 'View all', 'woo-gutenberg-products-block' )
```
This needs escaping. `esc_html__()`

danieldudzic replied:
Are you referring to `__( 'View all', 'woo-gutenberg-products-block' )`?

Could you elaborate further? I believe the escaping is sufficient here.

roykho replied:
Unless I am missing something, I am not seeing any escaping for that translator text.

danieldudzic replied:
Right, I believe translated strings are generally considered safe in terms of output (unless when getting echo'd directly in an HTML attribute or a JavaScript context). In this specific case, it's used within the HTML context, so it should be safe.

roykho replied:
On the contrary, you should never trust the translator. Please see this reference https://developer.wordpress.org/reference/functions/esc_html__/

danieldudzic replied:
Searching through our repo yields a lot of results of translated unescaped plain text strings. If these are considered unsafe, we should fix them.

I did some searching and found this very interesting thread on escaping every translated string: poqVs-c0T-p2

It is quite old though, so I wonder if anything has changed since then.

roykho replied:
> Searching through our repo yields a lot of results of translated unescaped plain text strings. If these are considered unsafe, we should fix them.

Are those JS or PHP? Perhaps they're escaped later in the code? It is good practice to escape as late as possible.

> I did some searching and found this very interesting thread on escaping every translated string: poqVs-c0T-p2
> It is quite old though, so I wonder if anything has changed since then.

It seems there are proponent for using it in that P2 and the opposition is stating that instead of escaping, we should trust the translators by vetting them. But there is no way you can be sure they can be trusted. This is especially true when people are using 3rd party translation plugins which if some malicious user suggests to the shop owner to add this string in the translator, then that is a attacking vector they will leverage. Remember not all translations come from the official channels like GlotPress...etc.

If you have a look in Woo core, you can see `esc_html__` being used throughout the codebase https://github.com/search?q=repo%3Awoocommerce%2Fwoocommerce%20esc_html__&type=code

But I will yield to your discretion.

danieldudzic replied:
Thanks @roykho, I really appreciate you taking the time to respond to this. I had a follow-up Slack discussion with @nerrad regarding escaping translated strings: p1697572533886259-slack-C02FL3X7KR6 - I'm happy to make the update :)

roykho commented on src/BlockTypes/ProductGalleryThumbnails.php:
```diff
@@ -67,22 +108,31 @@ protected function render( $attributes, $content, $block ) {
 
-						$processor = new \WP_HTML_Tag_Processor( $product_gallery_image_html );
+						// If not in dialog and it's the last thumbnail and the number of product gallery images is greater than the number of thumbnails settings output the View All markup.
+						if ( 'standard' === $mode && $thumbnails_count === $number_of_thumbnails && count( $product_gallery_images ) > $number_of_thumbnails ) {
```
This conditional is quite hard to read. I would suggest extracting them into their own little functions.

danieldudzic replied:
Thank you, you are right. I have moved both conditionals to separate functions.

roykho reviewed (changes requested):
Left some inline comments. I also think perhaps we should change the action name from `handleClick` to `openLightbox`. Because an action should describe what it is trying to do and not what the event is. Also in the future if we need to add more "click", events it would be confusing to already have an action that is named `handleClick` which you don't really know which it is referring to at a glance. I know these were added prior to this PR but just pointing it out.

danieldudzic committed:
Product Gallery Thumbnails: Abstract the View All conditions into separate functions for readability

danieldudzic commented:
> I also think perhaps we should change the action name from handleClick to openLightbox. Because an action should describe what it is trying to do and not what the event is. Also in the future if we need to add more "click", events it would be confusing to already have an action that is named handleClick which you don't really know which it is referring to at a glance. I know these were added prior to this PR but just pointing it out.

I wholeheartedly agree. I'd change it from `handleClick` to `openDialog` to be consistent with the lightbox naming.

danieldudzic committed:
Product Gallery Thumbnails: Add escaping to the View all plain text string

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic commented:
I have now opened a separate issue for the name and the placement of the `handleClick` action: https://github.com/woocommerce/woocommerce-blocks/issues/11311

roykho commented:
I am seeing failing e2e tests. Could you please see if those are regressions?

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic commented:
> I am seeing failing e2e tests. Could you please see if those are regressions?

@roykho It seems that the failing test is trying to test the Sale Badge but gets tripped up by a second Sale Badge in the Dialog Product Gallery. This seems unrelated to my PR.

Should I open a separate bug report?

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

roykho commented:
> It seems that the failing test is trying to test the Sale Badge but gets tripped up by a second Sale Badge in the Dialog Product Gallery. This seems unrelated to my PR.

Interesting. The overlay is already in trunk. So I am wondering why it isn't failing tests there?

danieldudzic commented:
> > It seems that the failing test is trying to test the Sale Badge but gets tripped up by a second Sale Badge in the Dialog Product Gallery. This seems unrelated to my PR.
> 
> Interesting. The overlay is already in trunk. So I am wondering why it isn't failing tests there?

The overlay? Sorry, I'm not sure I understand. This PR adds the 'View All' overlay.

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
E2E: Fix the Sale Badge and Single Product Template tests by selecting the first Sale Badge

danieldudzic committed:
Merge branch 'fix/11083-product-gallery-thumbnails-add-view-all' of https://github.com/woocommerce/woocommerce-blocks into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

roykho reviewed (approved):
Great job! Tested nicely as described and thanks for fixing the e2e tests.

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all

danieldudzic committed:
Merge branch 'trunk' into fix/11083-product-gallery-thumbnails-add-view-all
